File: cost_backend/db/schemas/task.py
"""异步任务作业 Schema。"""
from typing import Any

from pydantic import BaseModel, ConfigDict


class TaskJobResp(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    job_id: str
    task_type: str
    status: str
    progress: int = 0
    total: int = 0
    processed: int = 0
    result: dict[str, Any] | None = None
    error: str | None = None
    created_by: int | None = None
    created_at: str | None = None
    started_at: str | None = None
    finished_at: str | None = None

File: cost_web/src/api/task.ts
import request from './request'

export interface TaskJobVO {
  job_id: string
  task_type: string
  status: 'pending' | 'running' | 'success' | 'failed'
  progress: number
  total: number
  processed: number
  result: any
  error: string | null
  created_by: number | null
  created_at: string | null
  started_at: string | null
  finished_at: string | null
}

// 查询异步任务进度与结果
export function getTaskStatusApi(jobId: string) {
  return request.get<TaskJobVO>(`/task/${jobId}`)
}

// 下载异步任务产出的文件（导出类任务）
export async function downloadTaskApi(jobId: string, filename = 'export.csv') {
  const blob = (await request.get(`/task/${jobId}/download`, {
    responseType: 'blob',
  })) as unknown as Blob
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = filename
  document.body.appendChild(a)
  a.click()
  a.remove()
  URL.revokeObjectURL(url)
}
